File: packages/python/dirsql/cli/interpret/dispatch_extract.py
"""Single-request handler for the `dirsql interpret` helper.

Given a parsed request line and a name->table lookup, call the user's
``extract`` callback and return the NDJSON response payload the loop
will write back to stdout.
"""

from __future__ import annotations

from typing import Any


def dispatch_extract(
    req: dict[str, Any],
    tables: dict[str, Any],
) -> dict[str, Any]:
    """Build the `{"type": "result", ...}` response for one extract request.

    - Unknown table name -> ``ok: false`` with a name-bearing error.
    - User's ``extract`` raising -> ``ok: false`` with ``str(exc)``.
    - Success -> ``ok: true`` with the row list the callback returned.

    ``req["id"]`` is echoed back unchanged so the caller can correlate
    request and response.
    """
    rid = req.get("id")
    name = req.get("table")
    path = req.get("path")
    table = tables.get(name)
    if table is None:
        return {
            "type": "result",
            "id": rid,
            "ok": False,
            "error": f"unknown table: {name!r}",
        }
    try:
        rows = table.extract(path)
    except Exception as exc:
        return {"type": "result", "id": rid, "ok": False, "error": str(exc)}
    return {"type": "result", "id": rid, "ok": True, "rows": rows}

File: packages/ts/src/cli/interpret/dispatch-extract.ts
// Single-request handler. Given a parsed extract request and the
// `name -> TableDef` map, call the user's extract callback (which may
// be sync or async) and build the NDJSON response payload the loop
// will write back.

import type { TableDef } from "../../index.js";
import { errMessage } from "./err-message.js";

export interface ExtractRequest {
  type?: string;
  id?: unknown;
  table?: string;
  path?: string;
}

export interface ResultMessage {
  type: "result";
  id: unknown;
  ok: boolean;
  rows?: unknown[];
  error?: string;
}

export async function dispatchExtract(
  req: ExtractRequest,
  tables: Map<string, TableDef>,
): Promise<ResultMessage> {
  const { id, table: name, path } = req;
  const table = name === undefined ? undefined : tables.get(name);
  if (!table) {
    return {
      type: "result",
      id,
      ok: false,
      error: `unknown table: ${JSON.stringify(name)}`,
    };
  }
  try {
    const rows = await table.extract(path as string);
    return { type: "result", id, ok: true, rows };
  } catch (e) {
    return { type: "result", id, ok: false, error: errMessage(e) };
  }
}
